source/simulator/nullspace.cc

```diff
@@ -19,22 +19,134 @@
 */
 
 
-#include <aspect/simulator.h>
 #include <aspect/global.h>
+#include <aspect/simulator.h>
+#include <aspect/geometry_model/sphere.h>
+#include <aspect/geometry_model/spherical_shell.h>
 
 #include <deal.II/lac/solver_gmres.h>
-
 #include <deal.II/lac/trilinos_solver.h>
 
 #include <deal.II/base/tensor_function.h>
 
 #include <deal.II/base/quadrature_lib.h>
 #include <deal.II/fe/fe_values.h>
 
+
 namespace aspect
 {
   namespace internal
   {
+    /**
+     * Add constraints to @p constraints for the active components
+     * specified by @p mask for the DoFs in the support point given by
+     * @p location assumed on the boundary of the domain.
+     *
+     * The DoFs are found by iterating over all faces on the boundary
+     * and computing the distance to @p location.
+     */
+    template <int dim, int spacedim=dim>
+    void
+    constrain_point_on_boundary(const DoFHandler<dim,spacedim> &dof_handler,
+                                const Mapping<dim> &mapping,
+                                const Point<dim> &location,
+                                const ComponentMask &mask,
+                                AffineConstraints<double> &constraints)
+    {
+      const auto &fe = dof_handler.get_fe();
+      const std::vector<Point<dim - 1>> &unit_support_points = fe.get_unit_face_support_points();
+      const Quadrature<dim - 1> quadrature(unit_support_points);
+      const unsigned int dofs_per_face = fe.dofs_per_face;
+      std::vector<types::global_dof_index> face_dofs(dofs_per_face);
+
+      FEFaceValues<dim, spacedim> fe_face_values(mapping,
+                                                 fe,
+                                                 quadrature,
+                                                 update_quadrature_points |
+                                                 update_normal_vectors);
+
+      for (const auto &cell : dof_handler.active_cell_iterators())
+        if (cell->is_ghost() || cell->is_locally_owned())
+          for (unsigned int face_no = 0;
+               face_no < GeometryInfo<dim>::faces_per_cell; ++face_no)
+            {
+              typename DoFHandler<dim, spacedim>::face_iterator face = cell->face(face_no);
+              face->get_dof_indices(face_dofs);
+              fe_face_values.reinit(cell, face_no);
+
+              for (unsigned int i = 0; i < face_dofs.size(); ++i)
+                {
+                  const Point<dim> position = fe_face_values.quadrature_point(i);
+                  if (position.distance(location) < 1e-6*cell->diameter())
```

Comment:
This check assumes the position can be found. Better check after the loop that it indeed was found.